main.js

```diff
@@ -254,9 +257,60 @@ define(function (require, exports, module) {
     }
   }
   
+  function onCodeChange (cm, change) {
+    sendLocalChange(cm, change)
+  }
+  
+  function onCodeChanges (cm, changes) {
+    sendLocalCaretsMoves(cm, changes)
+  }
+  
+  function onBeforeSelectionChange (cm, change) {
+    sendLocalSelect(cm, change)
+    sendLocalCaretsMoves(cm, change)
+  }
+  
+  function isSelection (range) {
+    return range.head.ch !== range.anchor.ch || range.head.line !== range.anchor.line
+  }
+  
+  function sendLocalCaretsMoves (cm, change) {
```

Comment:
Do you mind adding a few comments to this to explain how you are calculating the caret positions from the changes?